src/lib/comms/sol-netctl-impl-connman.c

```diff
@@ -0,0 +1,1133 @@
+            r = get_manager_properties(m);
+            SOL_INT_CHECK(r, < 0, r);
+        } else {
+            SOL_DBG("Ignored global manager property: %s", str);
+            r = sd_bus_message_skip(m, "v");
+            SOL_INT_CHECK(r, < 0, r);
+        }
+
+        r = sd_bus_message_exit_container(m);
+        SOL_INT_CHECK(r, < 0, r);
+    } while (1);
+
+end:
+    if (r == 0)
+        r = sd_bus_message_exit_container(m);
+    else
+        return r;
+
+    call_manager_monitor_callback();
+
+    return r;
+}
+
+static int
+dbus_connection_get_manager_properties(void)
+{
+    sd_bus *bus = sol_bus_client_get_bus(_ctx.connman);
+
+    SOL_NULL_CHECK(bus, -EINVAL);
+
+    return sd_bus_call_method_async(bus, &_ctx.manager_slot,
+        "net.connman", "/", "net.connman.Manager", "GetProperties",
+        _manager_properties_changed, &_ctx, NULL);
+}
+
+static int
+dbus_connection_get_service_properties(void)
+{
+    sd_bus *bus = sol_bus_client_get_bus(_ctx.connman);
+
+    SOL_NULL_CHECK(bus, -EINVAL);
+
+    return sd_bus_call_method_async(bus, &_ctx.service_slot,
+        "net.connman", "/", "net.connman.Manager", "GetServices",
+        _services_properties_changed, &_ctx, NULL);
+}
+
+SOL_API enum sol_connman_state
+sol_connman_get_state(void)
+{
+    return _ctx.connman_state;
+}
+
+static int
+_set_state_property_changed(sd_bus_message *reply, void *userdata,
+    sd_bus_error *ret_error)
+{
+    struct ctx *pending = userdata;
+
+    pending->state_slot = sd_bus_slot_unref(pending->state_slot);
+
+    return sol_bus_log_callback(reply, userdata, ret_error);
+}
+
+SOL_API int
+sol_connman_set_radios_offline(bool enabled)
+{
+    int r;
+    sd_bus *bus = sol_bus_client_get_bus(_ctx.connman);
+
+    SOL_NULL_CHECK(bus, -EINVAL);
+
+    r = sd_bus_call_method_async(bus, &_ctx.state_slot, "net.connman", "/",
+        "net.connman.Manager", "SetProperty", _set_state_property_changed,
+        &_ctx, "sv", "OfflineMode", "b", enabled);
+    SOL_INT_CHECK(r, < 0, r);
+
+    return 0;
+}
+
+SOL_API bool
+sol_connman_get_radios_offline(void)
+{
+    if (_ctx.connman_state != SOL_CONNMAN_STATE_OFFLINE)
+        return false;
+
+    return true;
+}
+
+static int
+_service_connect(sd_bus_message *reply, void *userdata,
+    sd_bus_error *ret_error)
+{
+    struct sol_connman_service *service = userdata;
+
+    SOL_NULL_CHECK(service, -EINVAL);
+
+    service->slot = sd_bus_slot_unref(service->slot);
+
+    if (sol_bus_log_callback(reply, userdata, ret_error) < 0)
+        call_error_monitor_callback(false);
+    else
+        call_error_monitor_callback(true);
+
+    return 0;
+}
+
+SOL_API int
+sol_connman_service_connect(struct sol_connman_service *service)
+{
+    sd_bus *bus = sol_bus_client_get_bus(_ctx.connman);
+
+    SOL_NULL_CHECK(bus, -EINVAL);
+    SOL_NULL_CHECK(service, -EINVAL);
+    SOL_NULL_CHECK(service->path, -EINVAL);
+
+    return sd_bus_call_method_async(bus, &service->slot, "net.connman", service->path,
```

Comment:
I will use the above option to modify it.(one option is to return -EBUSY or something like that if there is a pending call)